api/src/wfl/source.clj

```diff
@@ -249,21 +256,36 @@
 
 (defn ^:private find-and-snapshot-new-rows
   "Create and enqueue snapshots from new rows in the `source` dataset."
-  [{:keys [last_checked] :as source} utc-now]
-  (let [shards->jobs (->> [(timestamp-to-offsetdatetime last_checked) utc-now]
-                          (mapv #(.format % bigquery-datetime-format))
-                          (find-new-rows source)
-                          (create-snapshots source utc-now))]
-    (when (seq shards->jobs)
-      (write-snapshots-creation-jobs source utc-now shards->jobs)
-      (update-last-checked source utc-now))))
+  [{:keys [dataset table last_checked] :as source}
+   utc-now]
+  (let [[start end :as interval]
+        (mapv
+         #(.format % bigquery-datetime-format)
+         [(timestamp-to-offsetdatetime last_checked) utc-now])
+        shards->jobs
+        (create-snapshots source utc-now (find-new-rows source interval))]
+    (if (seq shards->jobs)
+      (do (log/infof "%s Snapshots created from new rows in %s.%s."
+                     (log-prefix source) (:name dataset) table)
+          (write-snapshots-creation-jobs source utc-now shards->jobs)
+          (update-last-checked source utc-now))
+      (log/debugf "%s No rows in %s.%s between [%s, %s]."
+                  (log-prefix source) (:name dataset) table start end))))
```

Comment:
I would log that you're going to do something and let the logging down the call stack give this detail. For example, ehen you update the source, log "updating my-source-type" or similar then log what it has done (or what's happened), not what it hasn't done - adding if-then-else for logging introduces too much clutter to make it worthwhile.

Reply:
Sounds good.  I felt uneasy about the readability sacrifice as well.

I thought it could be worthwhile in a future where WFL is running many workloads simultaneously, with their logs interleaved.  But if we reach that point, I hope we will also be thinking about splitting the log file!